staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -353,8 +374,31 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
 	a.cachedCreds = newCreds
 	// Only close all connections when TLS cert rotates. Token rotation doesn't
 	// need the extra noise.
-	if a.onRotate != nil && oldCreds != nil && !reflect.DeepEqual(oldCreds.cert, a.cachedCreds.cert) {
-		a.onRotate()
+	if oldCreds != nil && !reflect.DeepEqual(oldCreds.cert, a.cachedCreds.cert) {
+		// Can be nil if the exec auth plugin only returned token auth.
+		if oldCreds.cert != nil && oldCreds.cert.Leaf != nil {
+			metrics.ClientCertRotationAge.Observe(time.Now().Sub(oldCreds.cert.Leaf.NotBefore))
+		}
+		if a.onRotate != nil {
+			a.onRotate()
+		}
+	}
+	// Update the nearest client cert expiration from cache as it likely changed.
+	metrics.ClientCertExpiration.Set(globalCache.oldestExpiration())
```

Comment:
should this be inside the `if` statement so it only runs when the current cert actually changed??

Reply:
Updated the function `oldestExpiration()` to -> `shortestTTL()` but both implementations are safe and intended to log in all situations.

The function `shortestTTL()` returns the following in each of the cases:

- *no cert -> no cert*: <nil>
- *no cert -> cert*: duration of shortest TTL
- *cert -> cert*: duration of (probably longer) shortest TTL
- *cert -> no cert*: <nil>